pkg/clustermanager/provider-cluster/provider_cluster.go

```diff
@@ -0,0 +1,156 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package providercluster
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog/v2"
+
+	clusterproviderclient "github.com/kcp-dev/edge-mc/cluster-provider-client"
+	clusterprovider "github.com/kcp-dev/edge-mc/cluster-provider-client/cluster"
+	lcv1alpha1apis "github.com/kcp-dev/edge-mc/pkg/apis/logicalcluster/v1alpha1"
+	edgeclient "github.com/kcp-dev/edge-mc/pkg/client/clientset/versioned"
+)
+
+var (
+	providerList map[string]clusterproviderclient.ProviderClient
+	lock         sync.Mutex
+)
+
+func New() {
+	lock.Lock()
+	defer lock.Unlock()
+	providerList = make(map[string]clusterproviderclient.ProviderClient)
+}
+
+func GetProvider(ctx context.Context,
+	clusterclientset edgeclient.Interface,
+	providerName string,
+	providerType lcv1alpha1apis.ClusterProviderType) (clusterproviderclient.ProviderClient, error) {
+	lock.Lock()
+	defer lock.Unlock()
+
+	provider, exists := providerList[providerName]
+	if !exists {
+		newProvider, err := clusterproviderclient.NewProvider(ctx, providerName, providerType)
+		if err != nil {
+			return nil, err
+		}
+		providerList[providerName] = newProvider
+		w, err := newProvider.Watch()
+		if err != nil {
+			return nil, err
+		}
+		go ProcessProviderWatchEvents(ctx, w, clusterclientset, providerName)
+		provider = newProvider
+	}
+	return provider, nil
+}
+
+func CreateProvider(ctx context.Context,
+	providerName string,
+	providerType lcv1alpha1apis.ClusterProviderType) (clusterproviderclient.ProviderClient, error) {
+	lock.Lock()
+	defer lock.Unlock()
+
+	_, exists := providerList[providerName]
+	if exists {
+		err := errors.New("provider already in the list")
+		return nil, err
+	}
+	newProvider, err := clusterproviderclient.NewProvider(ctx, providerName, providerType)
+	if err != nil {
+		return nil, err
+	}
+	providerList[providerName] = newProvider
+	return newProvider, nil
+}
+
+func StartDiscovery(ctx context.Context,
+	provider clusterproviderclient.ProviderClient,
+	clusterclientset edgeclient.Interface,
+	providerName string) error {
+	w, err := provider.Watch()
+	if err != nil {
+		return err
+	}
+	go ProcessProviderWatchEvents(ctx, w, clusterclientset, providerName)
+	return nil
+}
+
+func ProcessProviderWatchEvents(ctx context.Context, w clusterprovider.Watcher, clientset edgeclient.Interface, providerName string) {
+	logger := klog.FromContext(ctx)
+	for {
```

Comment:
you need a way to stop this loop from the manager

Reply:
Calling Watcher.Stop() will exit the loop.  Or it should.  I need to try it :)